edb/server/server.py

```diff
@@ -802,6 +808,91 @@ async def _introspect_dbs(self):
                 # information about them.
                 g.create_task(self._early_introspect_db(dbname))
 
+    async def get_patch_count(self, conn):
+        """Apply any un-applied patches to the database."""
+        num_patches = await conn.sql_fetch_val(
+            b'''
+                SELECT json::json from edgedbinstdata.instdata
+                WHERE key = 'num_patches';
+            ''',
+        )
+        num_patches = json.loads(num_patches) if num_patches else 0
+        return num_patches
+
+    async def _prepare_patches(self, conn):
+        """Prepare all the patches"""
+        num_patches = await self.get_patch_count(conn)
+        schema = self._std_schema
+
+        patches = {}
+        patch_list = list(enumerate(pg_patches.PATCHES))
+        for num, (kind, patch) in patch_list[num_patches:]:
+            from . import bootstrap
+            sql, syssql, schema = bootstrap.prepare_patch(
+                num, kind, patch, schema, self._refl_schema,
+                self._schema_class_layout, self.get_backend_runtime_params())
+
+            patches[num] = (sql, syssql, schema)
+
+        return patches
+
+    async def _maybe_apply_patches(self, dbname, conn, patches, sys=False):
+        """Apply any un-applied patches to the database."""
+        num_patches = await self.get_patch_count(conn)
+        for num, (sql, syssql, _) in patches.items():
+            if num_patches <= num:
+                if sys:
+                    sql += syssql
+                logger.info("applying patch %d to database '%s'", num, dbname)
+                sql = tuple(x.encode('utf-8') for x in sql)
+                await conn.sql_fetch(sql)
+
+    async def _maybe_patch_db(self, dbname, patches):
+        logger.info("applying patches to database '%s'", dbname)
+
+        conn = await self.acquire_pgcon(dbname)
```

Comment:
`_use_pgcon`?

Reply:
these aren't the sys conn; we only use acquire_pgcon like this once, so I haven't added an abstraction